### ISSUE TITLE
How do you balance the offset loss?

### ISSUE THREAD

**Issue by Epiphqny** [xingyizhou/CenterTrack#55]
Hello, thanks for your work. I adopted offset loss in my tracking model, but the loss is so large and disturb other losses seriously, I want to ask have you encountered the loss balance problem, and could you provide the training log for reference to see the weights of different losses?

**Comment by xingyizhou** [comment]
That's a good question. The scale of the offset loss is affected by many hyper-parameters: the loss weight, the max-allowed frame distance in the video, and the scale/shift-augmentation factor for static image training. I would suggest setting small weights for those at first. E.g., max frame dist 2, shift 0.01, loss weight 0.1 (it would make the performance worse, at least on MOT). 
Unfortunately, my log is split into different files due to interpretation in training and is currently too hard to collect them. Let me know if you have further questions.

**Comment by Epiphqny** [comment]
@xingyizhou Thanks for your quick reply.I set the max frame dist to 1, and added the offset branch on the FCOS detector, the offset loss still disturb the classification loss seriously. Only when setting the weight of offset loss to 0.01 both the losses degrade, but i think the offset branch can not learn in this setting. Have you tried the offset branch in other one stage detector? that branch is so hard to train. Or can there be any modified L1 loss to balance the problem?

**Comment by xingyizhou** [comment]
Clearly I haven't tried tracking with FCOS. Here are considerations when applying it to an FPN-based detector (some of them may be trivial): 

  - How do you prepare the prior heatmap. A single one or multi-scale?
  - How do you assign the target FPN level? What do you do if the current center and the previous center are in different FPN levels? How do you normalize the scale across FPN level?
  - There are 3x3 positive cells in FCOS. Do you want them all to point to the previous center or do you build a correspondence between them?

  I don't have a good modified L1 loss in mind. If you wait (a lot) longer, we will release a FPN-version CenterNet. This might make things easier. Good luck :)

**Comment by Epiphqny** [comment]
Thanks for your answer, that's very useful. I implemented a simplest version with only:

- use centerness as heatmap, a single one

- haven't consider the FPN level now, used the default bbox assignment method.

- firstly we built a correspondence of each pos cell to previous center, the improved version builds the offset target of each pos as the offset of the centers.

Also, i found a bug in my pervious implementation which cause the balance problem, i forgot to close the flip in data augmentation, which cause the offset very large. After fixing it, the loss is normal. But the result still have some problem, i will improve it these days.

**Comment by xingyizhou** [comment]
If you visualize the centerness in FCOS, it doesn't look like a heatmap. The bounding box boundary still has a value of 0.5. I do have tried to replace the sqrt-based centerness in FCOS with a gaussian heatmap, it seems to give a 0.1mAP improvement (at least not worse).

**Comment by Epiphqny** [comment]
Yes the centerness values are totally different from heatmap values, but previously I thought they are of the same purpose for predicting object center peaks. But after several days debugging and tuning of the model, the fcos with offset branch still works badly. I will try different methods and see if they works. Perhaps that multiple points being responsible for estimating the identity of the same object may cause ambiguities for the network, which is a big difference between the CenterNet and FCOS, and makes the FCOS not suitable for tracking.

**Comment by xingyizhou** [comment]
Closing it for now. Feel free to reopen and continue the discussion.
